data_prep/generate_annual_averages.py

```diff
@@ -0,0 +1,191 @@
+    nc.variables[variable].setncattr('long_name', new_long_name)
+
+def has_valid_time_bounds(bounds_var):
+    '''
+    Verifies if the input NetCDF has a valid, and populated, time_bnds variable
+    '''
+    # bounds_var might be a np.masked_array, and if any elements are masked, it's bad
+    if (hasattr(bounds_var, 'mask') and (np.ma.is_masked(bounds_var[:]))):
+        return False
+    # or, bounds_var might be full of zeros, also bad
+    elif (not np.any(bounds_var)):
+        return False
+    else:
+        return True
+
+def update_annual_avg_file_time_metadata(nc, start_year=None):
+    '''
+    Converts time values of a newly produced annual average NetCDF file
+    to annual midpoints. Checks validity of the time bounds variable
+    (initially set by CDO), overwriting it if necessary, and assigns
+    correct units and calendar attributes.
+
+    time values are the calculated midpoint between <year n>-01-01 and
+    <year n+1>-01-01
+    time bounds are <year n>-01-01 00:00:00 and <year n+1>-01-01 00:00:00
+
+    Optional parameter: 
+        start_year - set to override the time variable's start year
+    '''
+    time_var = nc.variables['time']
+    units = time_var.units
+    calendar = time_var.calendar
+    new_times = []
+    bounds_var = nc.variables['time_bnds']
+    new_bounds = []
+    write_new_bounds = False
+
+    if start_year is None:
+        start_year = num2date(time_var[0], units, calendar).year
+    # we are (possibly) setting a new start_year, so units and bounds will change
+    else:
+        units = "days since {}-01-01 00:00:00".format(start_year)
+        time_var.units = units
+        write_new_bounds = True
+    end_year = start_year + time_var.shape[0]
+
+    bounds_var.units = units
+    bounds_var.calendar = calendar
+
+    if not has_valid_time_bounds(bounds_var):
+        write_new_bounds = True
+
+    for year in range(start_year, end_year):
+        days_to_mid = ((datetime(year, 1, 1) + relativedelta(years=1)) \
+            - datetime(year, 1, 1)).days/2
+        mid = datetime(year, 1, 1) + relativedelta(days=days_to_mid)
+        new_times.append(mid)
+        if write_new_bounds:
+            lower_bound = datetime(year, 1, 1)
+            upper_bound = datetime(year, 1, 1) + relativedelta(years=1)
+            new_bounds.append([lower_bound, upper_bound])
+
+    time_var[:] = date2num(new_times, units, calendar)
+    if write_new_bounds:
+        bounds_var[:] = date2num(new_bounds, units, calendar)
+
+def main(args):
+    vars = '|'.join(args.variables)
+    test_files = iter_matching(args.basedir, \
+        re.compile('.*({}).*(_rcp26|_rcp45|_rcp85|_historical_).*r1i1p1.*nc'\
+        .format(vars)))
+
+    if args.dry_run:
+        for f in test_files:
+            print(f)
+        sys.exit(0)
+
+    for fp in test_files:
+        log.info('Processing input file: {}'.format(fp))
+        file_ = Cmip5File(fp, freq='yr', mip_table='yr')
```

Comment:
I guess the intent was to conform to the CMIP5 standard for output file paths.  I guess a less rigid option would be to extract the metadata from the source NetCDF itself.